### ISSUE TITLE
Subnormal f64 to f32 cast is incorrect in CTFE

### ISSUE THREAD

**Issue by cbeuw** [rust-lang/rust#113407]
```rust
pub fn main() {
    let f = f64::from_bits(0x19873cc2) as f32;
    println!("{:x}", f.to_bits());
    println!("{}", f == 0.);
}
```

```console
$ rustc -Zmir-opt-level=2 -Copt-level=3 repro.rs && ./repro
1
false
$ ../miri/miri run repro.rs 2> /dev/null
1
false
$ rustc -Zmir-opt-level=1 -Copt-level=3 repro.rs && ./repro
0
true
```

It should print `0` and `true`, according to the equivalent C program: https://godbolt.org/z/7Gq51a1Yh

This is probably another apfloat bug

```console
$ rustc --version -v 
rustc 1.72.0-nightly (d9c13cd45 2023-07-05)
binary: rustc
commit-hash: d9c13cd4531649c2028a8384cb4d4e54f985380e
commit-date: 2023-07-05
host: aarch64-unknown-linux-gnu
release: 1.72.0-nightly
LLVM version: 16.0.5
```
(reproducible on x86 as well)

cc @eddyb @wesleywiser @RalfJung

**Comment by RalfJung** [comment]
This is strange: https://godbolt.org/z/14GxTsfoM

LLVM and Rust are both using apfloat for compile-time float computations, and yet the MIR opts produce 1 and LLVM produces 0. Is this something that was fixed in upstream apfloat but we don't have the fix yet?

**Comment by cbeuw** [comment]
Rust's apfloat is old and unupdated for a long while. For instance https://github.com/rust-lang/rust/issues/102403 was fixed upstream 6 years ago https://github.com/llvm/llvm-project/commit/f2c2851efe0505328d1f4074b313a2922bcd8f2c

**Comment by nikic** [comment]
The relevant fix is probably https://reviews.llvm.org/D127140?

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/113416 should work around this.

**Comment by eddyb** [comment]
While trying to trigger a different issue involving float<->float conversions, I believe [my `rustc_apfloat` fuzzer](https://github.com/LykenSol/rustc_apfloat#fuzzing) (after commit https://github.com/LykenSol/rustc_apfloat/commit/439177ac1c8ad07378032109539cea5759842428) reproduced this one instead:

![image](https://github.com/rust-lang/rust/assets/77424/27885e87-c1e8-44ea-85e4-15e8c9d55760)

(which is good confirmation that we can find such things. too bad about the FMA NaNs polluting fuzzer outputs)

**Comment by eddyb** [comment]
As per https://github.com/rust-lang/rust/issues/55993#issuecomment-1639054175, this is now fixed in `rust-lang/rustc_apfloat`.

**Comment by cbeuw** [comment]
This should be reopened until rustc starts using the stand-alone crate

**Comment by eddyb** [comment]
Uh oh I did not do that, GH must've gotten overzealous cross-repo.

**Comment by RalfJung** [comment]
You said "fixes  #113407" in that PR, GH did the rest. ;)
